ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala

```diff
@@ -0,0 +1,221 @@
+    connectionSource: DataSource with AutoCloseable,
+)(implicit executionContext: ExecutionContext, materializer: Materializer)
+    extends LedgerWriter
+    with LedgerReader
+    with AutoCloseable {
+
+  private val engine = Engine()
+
+  private val dispatcher: Dispatcher[Index] =
+    Dispatcher(
+      "sql-participant-state",
+      zeroIndex = FirstIndex,
+      headAtInitialization = FirstIndex,
+    )
+
+  private val randomNumberGenerator = new Random()
+
+  // TODO: implement
+  override def checkHealth(): HealthStatus = Healthy
+
+  override def close(): Unit = {
+    dispatcher.close()
+    connectionSource.close()
+  }
+
+  override def retrieveLedgerId(): LedgerId = ledgerId
+
+  override def events(offset: Option[Offset]): Source[LedgerRecord, NotUsed] =
+    dispatcher
+      .startingAt(
+        offset.getOrElse(FirstOffset).components.head,
+        RangeSource((start, end) =>
+          withConnection { implicit connection =>
+            Future(Source(queries.selectFromLog(start, end)))
+        })
+      )
+      .map { case (_, record) => record }
+
+  override def commit(correlationId: String, envelope: Array[Byte]): Future[SubmissionResult] =
+    withConnection { implicit connection =>
+      Future {
+        val submission = Envelope
+          .openSubmission(envelope)
+          .getOrElse(throw new IllegalArgumentException("Not a valid submission in envelope"))
+        val stateInputKeys: Set[DamlStateKey] = submission.getInputDamlStateList.asScala.toSet
+        val stateInputs = readState(stateInputKeys)
+        val entryId = allocateEntryId()
+        val (logEntry, stateUpdates) = KeyValueCommitting.processSubmission(
+          engine,
+          entryId,
+          currentRecordTime(),
+          LedgerReader.DefaultTimeModel,
+          submission,
+          participantId,
+          stateInputs,
+        )
+        verifyStateUpdatesAgainstPreDeclaredOutputs(stateUpdates, entryId, submission)
+        val newHead = appendLog(entryId, Envelope.enclose(logEntry))
+        queries.updateState(stateUpdates)
+        dispatcher.signalNewHead(newHead)
+        SubmissionResult.Acknowledged
+      }
+    }
+
+  private def verifyStateUpdatesAgainstPreDeclaredOutputs(
+      actualStateUpdates: Map[DamlStateKey, DamlStateValue],
+      entryId: DamlLogEntryId,
+      submission: DamlSubmission,
+  ): Unit = {
+    val expectedStateUpdates = KeyValueCommitting.submissionOutputs(entryId, submission)
+    if (!(actualStateUpdates.keySet subsetOf expectedStateUpdates)) {
+      val unaccountedKeys = actualStateUpdates.keySet diff expectedStateUpdates
+      sys.error(
+        s"CommitActor: State updates not a subset of expected updates! Keys [$unaccountedKeys] are unaccounted for!")
+    }
+  }
+
+  private def currentRecordTime(): Timestamp =
+    Timestamp.assertFromInstant(Clock.systemUTC().instant())
+
+  private def allocateEntryId(): DamlLogEntryId = {
+    val nonce: Array[Byte] = Array.ofDim(8)
```

Comment:
kvutils should probably provide a canonical way to allocate a random entry id with enough entropy. This way of allocating doesn't provide enough entropy. For the in-memory and SQL implementations we might want to consider just allocating the ids sequentially.

kvutils also doesn't currently declare the outputs as inputs and hence does not check for conflicts on outputs itself (e.g. we might end up overwriting an existing log entry!). We might want to consider doing this check in kvutils itself. FYI @miklos-da.

Reply:
It does have the downside that we end up doing more fetches for keys that we expect not to exist and depending on the ledger we might have other options for detecting the conflict (e.g. allocating entry ids in a conflict free way).

Reply:
You're right about just using sequential IDs here. Will fix at some point.

Reply:
Regarding the checks, I think it's a good idea. Perhaps the implementation can decide whether to add it to the list of inputs or not.